### PR TITLE
elixir: Add textobjects queries

### PR DESCRIPTION
**Pull request by baldwindavid**
Release Notes:

- N/A

Adds textobjects queries for Elixir.

These queries were originally pulled directly from the [nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects) repo with [this textobjects.scm](https://raw.githubusercontent.com/nvim-treesitter/nvim-treesitter-textobjects/refs/heads/master/queries/elixir/textobjects.scm) file, but have been heavily edited for Zed.

### CODE REVIEW

**Comment by ConradIrwin** [comment]
@baldwindavid thanks for this!

Unfortunately Zed and nvim work a bit differently (though I was initially trying to build a compatible implementation with either nvim or helix, neither quite worked for our context).

To update the queries, you'll need to remove use of `make-range` etc. (compare https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/zig/textobjects.scm with [extensions/zig/languages/zig/textobjects.scm](https://github.com/zed-industries/zed/blob/c42c1b9e5c07e990bb99098e0eec16a75ff62811/extensions/zig/languages/zig/textobjects.scm#L1) for example.

Zed also does not support most of the object types that nvim does; currently it's just function/class/comment, so you can remove most of them

**Comment by baldwindavid** [comment]
Thanks, @ConradIrwin . Okay, this has been mostly gutted and ready for another look.

The only surprising thing I notice is that visual select around functions (`vaf`) selects the entire lines rather than just the function which is problematic for inline/anonymous functions. It appears it is that way in other languages too though and at least `daf/caf` work as expected.

**Comment by ConradIrwin** [comment]
Thanks! 

I made `vaf` target Visual line, as that seemed more consistent with vim. Does `vif` do what you want? 

For languages where it was possible to distinguish I tried to make the method queries only select "top level" methods; because the same definition is used for vim's `[m` etc. motions; I didn't want them stopping at closures (but best effort, because many languages don't support a strong distinction). In that scenario line-mode made a little more sense, but for languages with more succinct syntax it's a bit of a meh either way... Happy to change it if you think we should.

**Comment by baldwindavid** [comment]
I'll strictly speak about Elixir here and, as a blanket statement, I'd just say I don't want to negatively impact other languages. I think you're already ahead of me here, but want to make sure I'm clear about what I'm seeing in case there's any discrepancy.

Yes, `vif` does exactly what I'd expect which is selecting inside the anonymous function (i.e. between `->` and `end`).

2025-01-07T171209@2x

I would have originally expected `vaf` to select just the anonymous function (i.e. `fn _ -> {nil, value} end`) but it selects all lines where the anonymous function is on.

2025-01-07T171303@2x

> I tried to make the method queries only select "top level" methods

As of now `]m` and `[m` only hit top-level `def` functions as I would expect. If you're saying changing the behavior of `v` in the case of `vaf` would change it to such that anonymous functions would be visited by `]m` then, yeah, I'd probably rather that not happen.

More generally, I guess I just don't expect `v` to be line-wise because I think of it as always character-wise and `V` as line-wise. It's very possible that is an assumption that doesn't hold true in vim.

In the end this probably doesn't matter to me all _that much_ because I'm likely to be using `caf`, `yaf`, `daf` rather than `vaf`, outside maybe selecting an anonymous function and then walking up the node tree from there. It was mostly just noticeable when testing this because I thought I had broken something. `caf` (screenshot below), `yaf`, and `daf` all operate on the expected characters rather than the entire line changes so all good there.

2025-01-07T171419@2x

So basically, I'd lean to however you think makes the most sense across languages.

**Comment by ConradIrwin** [comment]
Makes sense! I'll leave it how it is for now, but adjust if/when we get more feedback/experience using it.
